daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala

```diff
@@ -483,41 +481,82 @@ private[validation] object Typing {
-          handleLookup(ctx, interface.lookupInterface(impl.interfaceId))
-
-        requires
-          .filterNot(impls.contains)
-          .foreach(required => throw EMissingRequiredInterface(ctx, tplTcon, iface, required))
+      requires
+        .filterNot(required =>
+          interface.lookupInterfaceCoImplements(tplTcon, required).isRight ||
+            interface.lookupTemplateImplements(tplTcon, required).isRight
+        )
+        .foreach(required => throw EMissingRequiredInterface(ctx, tplTcon, ifaceTcon, required))
 
-        methods.values.foreach { (method: InterfaceMethod) =>
-          if (!impl.methods.contains(method.name))
-            throw EMissingInterfaceMethod(ctx, tplTcon, impl.interfaceId, method.name)
-        }
-        impl.methods.values.foreach { (tplMethod: TemplateImplementsMethod) =>
-          methods.get(tplMethod.name) match {
-            case None =>
-              throw EUnknownInterfaceMethod(ctx, tplTcon, impl.interfaceId, tplMethod.name)
-            case Some(method) =>
-              checkExpr(tplMethod.value, method.returnType)
-          }
+      methods.values.foreach { (method: InterfaceMethod) =>
+        if (!implMethods.exists { case (name, _) => name == method.name })
+          throw EMissingInterfaceMethod(ctx, tplTcon, ifaceTcon, method.name)
+      }
+      implMethods.foreach { case (name, value) =>
+        methods.get(name) match {
+          case None =>
+            throw EUnknownInterfaceMethod(ctx, tplTcon, ifaceTcon, name)
+          case Some(method) =>
+            checkExpr(value, method.returnType)
         }
       }
     }
 
-    // TODO (MA): https://github.com/digital-asset/daml/issues/14047
-    @nowarn("cat=unused&msg=parameter value ifaceTcon in method")
-    @nowarn("cat=unused&msg=parameter value coImpls in method")
+    private def checkIfaceImplementation(
+        tplTcon: TypeConName,
+        impl: TemplateImplements,
+    ): Unit = {
+      val ifaceTcon = impl.interfaceId
+      interface
+        .lookupInterfaceCoImplements(tplTcon, ifaceTcon)
+        .foreach(_ => throw EConflictingImplementsCoImplements(ctx, tplTcon, ifaceTcon))
+      checkGenImplementation(
+        tplTcon,
+        ifaceTcon,
+        impl.methods.values.map(TemplateImplementsMethod.unapply(_).value).toList,
+      )
+    }
+
+    private def checkIfaceImplementations(
+        tplTcon: TypeConName,
+        impls: Map[TypeConName, TemplateImplements],
+    ): Unit =
+      impls.values.foreach(checkIfaceImplementation(tplTcon, _))
+
+    private def checkIfaceCoImplementation(
+        ifaceTcon: TypeConName,
+        param: ExprVarName,
+        coImpl: InterfaceCoImplements,
+    ): Unit = {
+      val tplTcon = coImpl.templateId
+
+      // NOTE (MA): we don't need to look for conflicting TemplateImplements since
+      // checkIfaceImplementation already does the other side of the check.
+
+      // Note (MA): we need to introduce `param : TTyCon(tplTcon)` to shadow
```

Comment:
I'm not super happy about this, but otherwise `param` (99% of the time `this`) would have the type of the interface in a `coimplements` and the type of the template in an `implements` and I don't like that asymmetry.

Reply:
why do we need to shadow the parameter? I would have expected that we just use a different env that doesn’t have `param` bound at all and then bind it here. Ofc the two end up being equivalent but it seems cleaner to use distinct envs than to try to get to this one by shadowing.

Reply:
@cocreature sounds resonable.

Reply:
I'll address this in a future PR